fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipIndexManager.java

```diff
@@ -0,0 +1,602 @@
+    private static final String COMMIT_ADDS =
+            "INSERT INTO membership" +
+            " (subject_id, property, object_id, source_id, start_time, end_time)" +
+            " SELECT subject_id, property, object_id, source_id, start_time, end_time" +
+            " FROM membership_tx_operations mto" +
+            " WHERE mto.tx_id = :txId" +
+                " AND mto.operation = :addOp" +
+                " AND NOT EXISTS (" +
+                    " SELECT TRUE" +
+                    " FROM membership m" +
+                    " WHERE m.source_id = mto.source_id" +
+                        " AND m.subject_id = mto.subject_id" +
+                        " AND m.property = mto.property" +
+                        " AND m.object_id = mto.object_id" +
+                        " AND m.start_time = mto.start_time" +
+                        " AND m.end_time = mto.end_time" +
+                " )";
+
+    private static final String DELETE_TRANSACTION =
+            "DELETE FROM membership_tx_operations" +
+            " WHERE tx_id = :txId";
+
+    private static final String TRUNCATE_MEMBERSHIP = "TRUNCATE TABLE membership";
+
+    private static final String TRUNCATE_MEMBERSHIP_TX = "TRUNCATE TABLE membership_tx_operations";
+
+    @Inject
+    private DataSource dataSource;
+
+    private NamedParameterJdbcTemplate jdbcTemplate;
+
+    private static final Map<DbPlatform, String> DDL_MAP = Map.of(
+            DbPlatform.MYSQL, "sql/mysql-membership.sql",
+            DbPlatform.H2, "sql/default-membership.sql",
+            DbPlatform.POSTGRESQL, "sql/default-membership.sql",
+            DbPlatform.MARIADB, "sql/default-membership.sql"
+    );
+
+    @PostConstruct
+    public void setUp() {
+        jdbcTemplate = new NamedParameterJdbcTemplate(getDataSource());
+
+        final var dbPlatform = DbPlatform.fromDataSource(dataSource);
+
+        Preconditions.checkArgument(DDL_MAP.containsKey(dbPlatform),
+                "Missing DDL mapping for %s", dbPlatform);
+
+        final var ddl = DDL_MAP.get(dbPlatform);
+        log.debug("Applying ddl: {}", ddl);
+        DatabasePopulatorUtils.execute(
+                new ResourceDatabasePopulator(new DefaultResourceLoader().getResource("classpath:" + ddl)),
+                dataSource);
+    }
+
+    /**
+     * End a membership entry, setting an end time if committed, or clearing from the current tx
+     * if it was newly added.
+     *
+     * @param txId transaction id
+     * @param sourceId ID of the direct/indirect container whose membership should be ended
+     * @param membership membership triple to end
+     * @param endTime the time the resource was deleted, generally its last modified
+     */
+    public void endMembership(final String txId,  final FedoraId sourceId, final Triple membership,
+            final Instant endTime) {
+        final Map<String, Object> parameterSource = Map.of(
+                "txId", txId,
```

Comment:
Perhaps the column names should be constants?

Reply:
The parameter names or the column names? I'm okay with constants for the parameter names

Reply:
The keys in the map